runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,517 @@
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Commit the transaction.
+     * For a transaction that only has write operations, this method will start and end
+     * a corfu transaction keeping the "critical" section small while batching up all updates.
+     * The commit returns successfully if the transaction was committed.
+     * Otherwise this throws a TransactionAbortedException with a cause.
+     * The cause and the caller's intent of the transaction can determine if this aborted
+     * Transaction can be retried.
+     * @return - address at which the commit of this transaction occurred.
+     */
+    public long commit() {
+        txBeginAndApply(null);
+        long commitAddress = Address.NON_ADDRESS;
+        if (TransactionalContext.isInTransaction()) {
+            try {
+                commitAddress = this.objectsView.TXEnd();
+            } catch (Exception ex) {
+                tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+                tablesInTxn.clear();
+                throw ex;
+            }
+            long timeElapsed = System.nanoTime() - txnStartTime;
+            switch (txnType) {
+                case readOnly:
+                    tablesInTxn.forEach(t -> t.getMetrics().readOnlyTxnTimes.set(timeElapsed));
+                    break;
+                case writeOnly:
+                    tablesInTxn.forEach(t -> t.getMetrics().writeOnlyTxnTimes.set(timeElapsed));
+                    break;
+                case writeAndReads:
+                    tablesInTxn.forEach(t -> t.getMetrics().readWriteTxnTimes.set(timeElapsed));
+                    break;
+                default:
+                    log.error("UNKNOWN TxnType!!");
+                    break;
+            }
+        }
+        tablesInTxn.clear();
+        return commitAddress;
+    }
+
+    /**
+     * Explicitly abort a transaction in case of an external failure
+     */
+    public void txAbort() {
+        operations.clear();
+        tablesInTxn.clear();
+        if (TransactionalContext.isInTransaction()) {
+            this.objectsView.TXAbort();
+            tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+        }
+    }
+
+    /**
+     * Cleanup the transaction resources.
+     * If invoked on transaction with just queries, record the time taken.
+     */
+    public void close() {
+        if (TransactionalContext.isInTransaction()) {
+            if (txnType == TxnType.readOnly) {
+                long timeElapsed = System.nanoTime() - txnStartTime;
+                tablesInTxn.forEach(t -> t.getMetrics().readOnlyTxnTimes.set(timeElapsed));
```

Comment:
I think we cannot check this condition here. Even if it's a read only txn, if an unexpected execption happens during txn and we come here to clean up resource, then the transaction is actually failed, so we should not set this metrics in that case, right?

Reply:
its a tradeoff - either we don't allow autocloseable on the queries or take the price of recording aborted read-only transaction times. I took the latter since autoclosable on reads is a useful feature requested by consumers.